text/0000-transaction-refactoring.md

```diff
@@ -0,0 +1,327 @@
+- `data` is the binary data that is returned from the last action of the `ActionReceipt`. Right now, it's empty for all actions except for function calls. For function calls the data is the result of the code execution. But in the future we might introduce non-contract state reads.
+
+Data should be stored at the same shard as the receiver's account, even if the receiver's account doesn't exist.
+
+### Examples
+
+#### Account Creation
+
+To create a new account we can create a new `Transaction`:
+
+```json
+{
+    originator_id: "vasya.near",
+    public_key: ...,
+    nonce: 42,
+    receiver_id: "vitalik.vasya.near",
+
+    action: [
+        {
+            create_account: {
+            }
+        },
+        {
+            transfer: {
+                amount: "19231293123"
+            }
+        },
+        {
+            deploy_contract: {
+                code: ...
+            }
+        },
+        {
+            add_key: {
+                public_key: ...,
+                access_key: ...
+            }
+        },
+        {
+            function_call: {
+                method_name: "init",
+                args: ...,
+                fee: "100010101",
+                amount: "0"
+            }
+        }
+    ]
+}
+```
+
+This transaction is sent from `vasya.near` signed with a `public_key`.
+The receiver is `vitalik.vasya.near`, which is a new account id.
+The transaction contains a batch of actions.
+First we create the account, then we transaction a few tokens on the new account, then we deploy code on the new account, add a new access key with some given public key, and as a final action initializing the deployed code by calling a method `init` with some arguments.
+
+For this transaction to work `vasya.near` needs to have enough balance to cover all actions at once.
+Every action has some associated transaction fee with it, plus `transfer` and `function_call` needs additional amounts and fees.
+
+Once we validated and subtracted the total fees+amounts from `vasya.near` account, this transaction will be transformed into a `Receipt`:
+
+```json
+{
+    sender_id: "vasya.near",
+    receiver_id: "vitalik.vasya.near",
+    receipt_id: ...,
+
+    action: {
+        originator_id: "vasya.near",
+        originator_public_key: ...,
+        refund_account_id: "vasya.near",
+        
+        output_data_id: null,
+        output_receiver_id: null,
+
+        input_data_id: [],
+        
+        action: [...]
+    }
+}
+```
+
+This receipt will be sent to `vitalik.vasya.near`'s shard to be executed.
+In case the `vitalik.vasya.near` account already exists, the execution will fail and some amount will be refunded back to `vasya.near` account with a new `ActionReceipt` with a single transfer action.
+BTW, if the refund fails (e.g. if `vasya.near` deletes his account in between), then the next `refund_account_id` will be `system`, which means the tokens are going to be burned and no receipts are generated.
```

Comment:
How do you delete an account?

Reply:
When you run out of rent. We don't support it yet, but it will be the case